drivers/pwm/rpm2pwm_mchp_xec.c

```diff
@@ -0,0 +1,248 @@
+	volatile uint8_t  STATUS;
+};
+
+struct fan_config {
+	const struct pwm_dt_spec *pwm;
+	uint8_t edges;
+};
+
+struct rpm2pwm_xec_config {
+	struct rpm2pwm_regs *const regs;
+	uint8_t pcr_idx;
+	uint8_t pcr_pos;
+	const struct pinctrl_dev_config *pcfg;
+	const struct fan_config *fan;
+};
+
+struct rpm2pwm_tach_xec_config {
+	struct device * const parent;
+};
+
+struct rpm2pwm_tach_xec_data {
+	uint16_t count;
+};
+
+struct rpm2pwm_xec_data {
+	uint32_t config;
+};
+
+int rpm2pwm_tach_xec_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	ARG_UNUSED(chan);
+
+	const struct rpm2pwm_tach_xec_config * const cfg = dev->config;
+	struct device * const parent = cfg->parent;
+	const struct rpm2pwm_xec_config * const parent_cfg = parent->config;
+	struct rpm2pwm_regs * const regs = parent_cfg->regs;
+
+	struct rpm2pwm_tach_xec_data * const data = dev->data;
+
+	data->count = 3932160 / (regs->TACH >> 3);
+
+	return 0;
+}
+
+static int rpm2pwm_tach_xec_channel_get(const struct device *dev,
+					enum sensor_channel chan,
+					struct sensor_value *val)
+{
+	struct rpm2pwm_tach_xec_data * const data = dev->data;
+
+	if (chan != SENSOR_CHAN_RPM) {
+		return -ENOTSUP;
+	}
+
+	val->val1 = data->count;
+	val->val2 = 0U;
+
+	return 0;
+}
+
+static const struct sensor_driver_api rpm2pwm_tach_xec_driver_api = {
+	.sample_fetch = rpm2pwm_tach_xec_sample_fetch,
+	.channel_get = rpm2pwm_tach_xec_channel_get,
+};
+
+static int rpm2pwm_tach_xec_init(const struct device *dev)
+{
+	const struct rpm2pwm_tach_xec_config *config = dev->config;
+
+	if (!device_is_ready(config->parent))
+		return -ENODEV;
+
+	return 0;
+}
+
+static int rpm2pwm_xec_set_cycles_internal(const struct device *dev, uint32_t channel,
+					   uint32_t period_count, uint32_t pulse_count,
+					   pwm_flags_t flags)
+{
+	const struct rpm2pwm_xec_config * const cfg = dev->config;
+	struct rpm2pwm_regs * const regs = cfg->regs;
+	const struct fan_config * const fan = cfg->fan;
+	uint8_t spinup;
+	uint16_t config;
+
+	spinup = regs->SPINUP;
+
+	/*
+	 * TODO: Figure out DTS flag options that are appropriate:
+	 * if (flags & RPM2PWM_XEC_FLAG_SPIN_NOKICK)
+	 */
+	spinup |= (1 << 5);
+
+	regs->SPINUP |= spinup;
+
+	config = (((fan->edges / 2) - 1) << 3);
+
+	regs->TARGET = (3932160 / pulse_count) << 3;
+
+	regs->CONFIG |= (config | 0x80);
+
+	return 0;
+}
+
+static int rpm2pwm_xec_set_cycles(const struct device *dev, uint32_t channel,
+				  uint32_t period_cycles, uint32_t pulse_cycles,
+				  pwm_flags_t flags)
+{
+	if (channel > 0) {
+		return -EIO;
+	}
+
+	rpm2pwm_xec_set_cycles_internal(dev, channel, period_cycles, pulse_cycles, flags);
+	return 0;
+}
+
+static int rpm2pwm_xec_get_cycles_per_sec(const struct device *dev,
+				      uint32_t channel, uint64_t *cycles)
+{
+	ARG_UNUSED(dev);
+
+	if (channel > 0) {
+		return -EIO;
+	}
+
+	if (cycles) {
+		/* User does not have to know about lowest clock,
+		 * the driver will select the most relevant one.
+		 */
+		*cycles = 32768;
```

Comment:
magic numbers, there is an API that returns cycles per second if that's the intention here.  sys_clock_hw_cycles_per_sec